Receiver: Fix quorum handling for all hashing algorithms

matej-g opened the pull request:
<!--
    Keep PR title verbose enough and add prefix telling
    about what components it touches e.g "query:" or ".*:"
-->

<!--
    Don't forget about CHANGELOG!

    Changelog entry format:
    - [#<PR-id>](<PR-URL>) Thanos <Component> ...

    <PR-id> Id of your pull request.
    <PR-URL> URL of your PR such as https://github.com/thanos-io/thanos/pull/<PR-id>
    <Component> Component affected by your changes such as Query, Store, Receive.
-->

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

This PR suggest a way to fix https://github.com/thanos-io/thanos/issues/5784 discovered with `ketama` hashing.

The presumption one replication = one request was true for `hashmod` hashing, however with `ketama` hashing this is no longer the case (same could be true for behavior for other algorithms in the future). There can now be requests being made to multiple endpoints as part of _single_ replicated request (i.e. one replication = multiple requests). As described in https://github.com/thanos-io/thanos/issues/5784, this renders the current quorum handling incorrect.

To fix quorum validation, I suggest to:
- _Introduce an additional way to validate quorum_ - instead of counting individual request, the focus should be on confirming if all requests that have been made _on behalf_ of a replica have succeeded (for simplification I refer to these as 'replica group request'). In this way we can confirm that if successful replica group requests >= success threshold, the quorum has been achieved. For example: for replication factor 3, we will end up with three replica group requests (1,2,3) - to guarantee quorum (2), at least two of these groups need to be written successfully.
- _Additionally validate quorum by counting endpoint failures_ - besides knowing that we need at least quorum of replica requests to succeed, we also know that due to hashing series each replica must end on a distinct node. This means that even if not all requests for any given replica group succeed, we still know quorum was reached if all request succeeded except for requests going to specific endpoint(s). The number of such failed endpoints can be determined based on quorum.

- This necessitates to send back the number of replica together with an error (or `nil` error in case of success), in order to make ti possible to track success of a replica group request and endpoint failures. This PR makes the necessary changes to errors channel in the forward fanout method.
- Some adjustments to the method determining write success were also made - when determining error for requests going to a particular replica, we need to either be able to find single cause for all errors or return single error signalizing that his replication failed.
- Another simplification included here is in when determining error cause - if the multi error includes only conflict errors, we do not need to care about the threshold, as we can assume other requests succeeded and we can directly return conflict.
- I also added read / write interactive tests (makes it easier to manually play with receiver changes)

![diagram drawio](https://user-images.githubusercontent.com/38492574/197976567-8a6cb90e-618b-49bf-ac46-db7f462d8350.png)

## Verification
- Added and extended unit / E2E tests for write error determination and quorum verifying
- Also ran manual tests and comparisons with / without changes, with replication, with / without outage

saswatamcode commented on pkg/receive/handler.go:
```diff
@@ -607,7 +607,15 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, wreqs map[e
+	ec := make(chan errWithReplicaFunc)
+
+	// replicaGroupReqs counts the number of requests that will be made on behalf
+	// on each replica. This is used to determine write success.
```
```suggestion
	// of each replica. This is used to determine write success.
```

saswatamcode commented on pkg/receive/handler.go:
```diff
@@ -747,34 +755,55 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, wreqs map[e
+					if success[i] >= replicaGroupReqs[i] {
+						// If enough replica groups succeed, we can finish early (quorum
+						// is guaranteed).
+						replicaGroupSuccess++
```
So each `replicaGroup` signifies all the requests made to replicate a batch of series to a `replica-endpoint`? And if all the requests within that `replicaGroup` succeed, we count that as a success towards ensuring quorum?

matej-g replied:
Yes, that's the gist :+1:, there is now also second way to verify quorum, I added more details in description.

saswatamcode commented on test/e2e/receive_test.go:
```diff
@@ -787,3 +631,92 @@ func TestReceive(t *testing.T) {
 	})
 }
+
+func runReplicationTest(t *testing.T, hashing receive.HashringAlgorithm, withOutage bool) {
```
💪🏻

saswatamcode commented on pkg/receive/handler.go:
```diff
@@ -747,34 +755,55 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, wreqs map[e
-	var success int
+	var (
+		// success counts how many requests per replica have succedeed.
+		success = make([]int, h.options.ReplicationFactor)
```
This change initializes some slices. Maybe it would be a good idea to benchmark and see changes to hot path?

matej-g replied:
I ran `BenchmarkHandlerReceiveHTTP`, I think we won't be able to avoid those few extra allocs but I think it should still be acceptable?

```
name                                                                                old time/op    new time/op    delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-12                         891µs ± 0%     942µs ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-12           1.07ms ± 0%    1.12ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-12                       9.42ms ± 0%    9.68ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-12          11.5ms ± 0%    11.2ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/OK-12                      39.7ms ± 0%    39.3ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/conflict_errors-12         49.7ms ± 0%    49.7ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-12                 85.7ms ± 0%   105.6ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-12    87.6ms ± 0%    89.9ms ± 0%   ~     (p=1.000 n=1+1)

name                                                                                old alloc/op   new alloc/op   delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-12                        1.15MB ± 0%    1.15MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-12           1.41MB ± 0%    1.41MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-12                       13.1MB ± 0%    13.1MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-12          15.5MB ± 0%    15.5MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/OK-12                      53.4MB ± 0%    53.4MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/conflict_errors-12         62.0MB ± 0%    62.0MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-12                  110MB ± 0%     110MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-12     110MB ± 0%     110MB ± 0%   ~     (p=1.000 n=1+1)

name                                                                                old allocs/op  new allocs/op  delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-12                         3.10k ± 0%     3.10k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-12            6.63k ± 0%     6.64k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-12                        30.3k ± 0%     30.3k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-12           65.2k ± 0%     65.2k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/OK-12                        121k ± 0%      121k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/conflict_errors-12           260k ± 0%      260k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-12                   89.0 ± 0%      93.0 ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-12       217 ± 0%       223 ± 0%   ~     (p=1.000 n=1+1)
```

saswatamcode replied:
Yup! Thanks!

matej-g replied:
With latest changes, still looking fine:

```
name                                                                                old time/op    new time/op    delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-12                        1.20ms ± 0%    0.87ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-12           1.25ms ± 0%    1.05ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-12                       10.6ms ± 0%     9.3ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-12          12.8ms ± 0%    10.8ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/OK-12                      46.3ms ± 0%    36.8ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/conflict_errors-12         55.5ms ± 0%    43.6ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-12                  102ms ± 0%      82ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-12     101ms ± 0%      86ms ± 0%   ~     (p=1.000 n=1+1)

name                                                                                old alloc/op   new alloc/op   delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-12                        1.15MB ± 0%    1.15MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-12           1.41MB ± 0%    1.41MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-12                       13.1MB ± 0%    13.1MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-12          15.5MB ± 0%    15.5MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/OK-12                      53.3MB ± 0%    53.2MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/conflict_errors-12         62.0MB ± 0%    62.0MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-12                  110MB ± 0%     110MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-12     110MB ± 0%     110MB ± 0%   ~     (p=1.000 n=1+1)

name                                                                                old allocs/op  new allocs/op  delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-12                         3.10k ± 0%     3.10k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-12            6.63k ± 0%     6.64k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-12                        30.3k ± 0%     30.3k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-12           65.2k ± 0%     65.2k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/OK-12                        121k ± 0%      121k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_20000_of_them/conflict_errors-12           260k ± 0%      260k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-12                   89.0 ± 0%      94.0 ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-12       218 ± 0%       224 ± 0%   ~     (p=1.000 n=1+1)

```

saswatamcode reviewed (commented):
Thanks for awesome work! 🌟 

Some minor comments.

PhilipGough commented:
@matej-g I am not familiar enough with this code path to determine correctness here or not, but determining quorum in this fashion makes sense to me for ketama

douglascamata commented on pkg/receive/handler.go:
```diff
@@ -747,34 +755,55 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, wreqs map[e
 
-	var success int
+	var (
+		// success counts how many requests per replica have succedeed.
```
Could we reflect this comment and all the others that follow into their respective variable name? This pattern of comment explaining "what" and not "why" is often a sign that something deserves a better name.

matej-g replied:
I tried to name it a bit, although I'm finding it challenging to try and find clear enough name that it would encapsulate the complexity of replication logic and be obvious without additional commentary. Suggestions welcome.

douglascamata commented on pkg/receive/handler.go:
```diff
@@ -976,6 +1011,16 @@ func determineWriteErrorCause(err error, threshold int) error {
 }
 
+// errWithReplicaFunc is a type to enable sending replica number and err over channel.
+type errWithReplicaFunc func() (uint64, error)
```
Why does this have to a function and can't be a `struct { uint64, error }`?

matej-g replied:
Does not have to, I found it simpler to "send" it via function, no strong preference here though.

douglascamata replied:
If we don't need a function, I think we should "downgrade" the type to a struct to be as simple as possible.

fpetkovski commented on pkg/receive/handler_test.go:
```diff
@@ -226,6 +226,15 @@ func TestDetermineWriteErrorCause(t *testing.T) {
 			threshold: 1,
 			exp:       errors.New("baz: 3 errors: 3 errors: qux; rpc error: code = AlreadyExists desc = conflict; rpc error: code = AlreadyExists desc = conflict; foo; bar"),
 		},
+		{
```
I wonder why this issue was not caught by tests. Do we run unit tests for Ketama?

matej-g replied:
We do have tests to verify quorum but they all run just with hashmod, I'm extending those as well to run with both algorithms.

fpetkovski replied:
Are you referring to the e2e tests that were extended? Can we also do the same for unit tests?

matej-g replied:
All relevant unit tests now have been expanded to run with both algorithms.

fpetkovski replied:
Out of curiosity, did tests fail before your fix?

matej-g replied:
Not really, which I believe was mainly because:
- With replication factor 3 and 3 nodes (which were all the test cases), even with ketama, it probably cannot be caught since the number of requests is small (but even on `main` expanding tests to simulate 6 node deployment, the failures start to show up
- We were running all tests with single timeseries, which is probably more of an edge case (now I run the tests with 50 series which should give us a good enough test data)
- The node addresses were randomized, I believe on `main` on ketama this was causing some incosistent results (I pushed fix for this now as well to generate the addresses in a predictable manner, no need for randomized addresses I believe)

douglascamata commented:
> There can now be requests being made to multiple endpoints as part of single replicated request (i.e. one replication = multiple requests).

I think that to avoid confusions we should clarify one fundamental difference between `ketama` and `hashmod`: what `ketama` really does is replicate each timeseries inside a request. This is independent of the amount of requests required to achieve it. Meanwhile, `hashmod` was replicating a single request.

For example, 2 timeseries with ketama using a replication factor 3 will require between 3 and 6 requests to be fully replicated. This will depend on their hash.

In fact, I am even curious whether our code can prevent 2 replicas of a given series from landing on the same node. Could be a low probability scenario, but on systems running millions of timeseries even a low probability might have a big impact. Do you know if this could happen, @fpetkovski or @matej-g?

fpetkovski commented:
This cannot happen anymore. It used to, but is now fixed by ensuring that the replica for each section in the ring is covered by a different node: https://github.com/thanos-io/thanos/blob/main/pkg/receive/hashring.go#L138-L154

matej-g commented:
Sorry for the delay folks.

I fixed a few more things, updated the description accordingly (I know there are quite few changes, if something's not clear or if I can make things easier to review by splitting / reorganizing this, please let me know).

I have also expanded all unit tests to verify quorum to run for both algorithms.

douglascamata commented on pkg/receive/handler.go:
```diff
@@ -747,34 +758,65 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, wreqs map[e
+	var (
+		// perReplicaSuccess is used to determine if enough requests succeeded for given replica number.
+		preReplicaSuccess = make([]int, h.options.ReplicationFactor)
+		// replicaMultiError contains all requests errors per replica.
+		replicaMultiError = make([]errutil.MultiError, h.options.ReplicationFactor)
```
So the variable could be named "replicaReqErrors", right?

matej-g replied:
Fixed :+1:

fpetkovski commented on pkg/receive/handler.go:
```diff
@@ -747,34 +758,65 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, wreqs map[e
-	var success int
+	var (
+		// perReplicaSuccess is used to determine if enough requests succeeded for given replica number.
+		preReplicaSuccess = make([]int, h.options.ReplicationFactor)
```
Shouldn't we be tracking per-shard success rate?

douglascamata replied:
What do you mean with "shard", @fpetkovski? Is is the same as the "replica group request" mentioned in the description of this pull request?

fpetkovski replied:
I think we might be talking about the same thing, but I still think this variable is very confusing. Requests get split into independent partitions (shards), and then each shard gets replicated. So I would expect us to track how many successful replications each shard has, e.g `perShardSuccessfulReplicas`.

matej-g replied:
I'm not sure I follow fully as well, but throwing in yet another term is even more confusing for me. Are we talking about the initial batching of requests in https://github.com/thanos-io/thanos/blob/main/pkg/receive/handler.go#L555 when you say 'shards'? Since most of the docs refer to these as 'batches', I'd either stick to this or refer to this as 'shards' everywhere (I don't have strong preference but I associate shards more with actual databases rather than requests).

Other than that you're right, but I named it 'per replica', since each index in the slice tracks success of requests going to given replica (for a _single_ shard / batch), but if it's not obvious enough I can prefix it as you suggested.

fpetkovski replied:
Calling it a batch makes sense as well, I agree that sharding is not mentioned anywhere else.

douglascamata replied:
If I'm not mistaken, in most of the places in this part of the code we could literally replace the world "replica" with "replicationIndex". This confuses me a lot.

matej-g replied:
Agreed @douglascamata, I think this where lot of the confusion originates, as @fpetkovski also mentions, this creates impression that replica = receiver node, but it's more of a logical replica (since parts of that replica will end up on different nodes). I'd be happy to rename this, I'll also try to improve the other variables and get rid of the "group request" naming.

douglascamata replied:
Amaaaaaazing, @matej-g!

bwplotka reviewed (commented):
Nice work, thanks! I think I need to spend more on understanding those two success modes. Something does not add up to me... but maybe it's late here (:

bwplotka commented:
Furthermore to me, e2e test `receive_ketama_outage` is unconclusive. It passes with:

* Old Thanos version before this PR
* With Thanos code from your PR
* With second success mode disabled 🙈 

![image](https://user-images.githubusercontent.com/6950331/199317271-d9c1feb0-1e7c-46d4-9ce0-fbed0328dd23.png)

bwplotka reviewed (changes requested):
Amazing work, but I feel there is a chance to push the https://github.com/thanos-io/thanos/issues/5807 idea further and simplify the whole flow tremendously. 🤔 Let's investigate it more, will try to help, plus we can talk on PromCon!

fpetkovski commented:
I agree that the current way replication works is not ideal, but if this is an issue we should probably prioritize fixing it. So unless the idea from #5807 is easy to implement, I don't think we should delay the fix until we have a pretty solution.

matej-g commented:
@fpetkovski we had finally some F2F time yesterday on this and we're considering if there's better alternative to this, but I'd agree with prioritizing the fix. I'll try to have this resolved by the end of this week.

fpetkovski commented on pkg/receive/handler.go:
```diff
@@ -933,17 +1001,31 @@ func (a expectedErrors) Len() int           { return len(a) }
+// It will inspect the error's cause if the error is a MultiError. It will return cause of each contained error but will not traverse any deeper.
+// If no cause can be determined, the original error is returned. When forReplication is true,
+// it will return failed replication error on any unknown error cause.
+func determineWriteErrorCause(err error, threshold int, forReplication bool) error {
```
Is this fixing a separate issue?

fpetkovski commented:
I tried solving the issue independently to see if we converge to the same solution: https://github.com/thanos-io/thanos/pull/5910

I think we're pretty much aligned, with the exception of sending a function over a channel which I found a bit confusing 😄

matej-g commented:
> I tried solving the issue independently to see if we converge to the same solution: #5910

Thanks, I took a look, it makes sense to me :+1:.

> I think we're pretty much aligned, with the exception of sending a function over a channel which I found a bit confusing smile I also still cannot wrap my head around counting successes by replica, maybe the math works out but I cannot do it easily in my head. If we count successes by batch instead, we can simply calculate quorum on each individual batch separately and fail if any of those quorums are not met.

For the sending function over func, I think @douglascamata also pointed this out, and I don't have a strong stance on that, we can send a struct as well, but I thought just to be able to send two multiple independent values this a common solution (but perhaps not).

As for the replication success counting, yes, I think we're on the similar page, but in #5910 you do it for each batch vs. here I try to achieve it globally, but your way seems less complex.

matej-g commented:
Closing as this has been superseded by https://github.com/thanos-io/thanos/pull/5910

douglascamata commented:
@matej-g I think you forgot to press the "Close" button.

fpetkovski commented:
Thanks @douglascamata for keeping track of this. Since @matej-g is away, let's close the issue for now. It's also out of sync with main. We can always ask for forgiveness from Matej if he objects :)
